test/test_resonator_spectroscopy.py

```diff
@@ -0,0 +1,93 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Spectroscopy tests for resonator spectroscop experiment."""
+
+from test.base import QiskitExperimentsTestCase
+from typing import Tuple
+import numpy as np
+
+from qiskit import QuantumCircuit
+
+from qiskit_experiments.library import ResonatorSpectroscopy
+from qiskit_experiments.test.mock_iq_backend import MockIQBackend
+
+
+class ResonatorSpectroscopyBackend(MockIQBackend):
+    """A simple and primitive backend to test spectroscopy experiments."""
+
+    def __init__(
+        self,
+        line_width: float = 2e6,
+        freq_offset: float = 0.0,
+        iq_cluster_centers: Tuple[float, float, float, float] = (-1.0, 0.0, 0.0, 0.0),
+        iq_cluster_width: float = 0.2,
+    ):
+        """Initialize the spectroscopy backend."""
+
+        super().__init__(iq_cluster_centers, iq_cluster_width)
+
+        self._linewidth = line_width
+        self._freq_offset = freq_offset
+        self._configuration.timing_constraints = {"granularity": 16}
+
+    def _compute_probability(self, circuit: QuantumCircuit) -> float:
+        """Returns the probability based on the frequency."""
+        freq_shift = next(iter(circuit.calibrations["measure"].values())).blocks[0].frequency
+        delta_freq = freq_shift - self._freq_offset
+
+        return np.abs(1 / (1 + 2.0j * delta_freq / self._linewidth))
+
+    def _iq_phase(self, circuit: QuantumCircuit) -> float:
+        """Add a phase to the IQ point depending on how far we are from the resonance.
+
+        This will cause the IQ points to rotate around in the IQ plane when we approach the
+        resonance which introduces and extra complication that the data processor needs to
+        properly handle.
+        """
+        freq_shift = next(iter(circuit.calibrations["measure"].values())).blocks[0].frequency
+        delta_freq = freq_shift - self._freq_offset
+
+        return delta_freq / self._linewidth
+
+
+class TestResonatorSpectroscopy(QiskitExperimentsTestCase):
+    """Tests for the resonator spectroscopy experiment."""
+
+    def test_end_to_end(self):
```

Comment:
Can you add more values for `res_freq` with ddt? Based on the background we needed to change fit function, it would be great if we can validate if the function is really robust.

Reply:
Done here: 77e8fb4